functions/sendgrid/index.js

```diff
@@ -13,101 +13,345 @@
-        return context.success('Email sent!');
+      if (response.statusCode < 200 || response.statusCode >= 400) {
+        console.error(response);
+      } else {
+        console.log('Email sent to: ' + req.body.to);
+      }
+
+      // Forward the response back to the requester
+      res.status(response.statusCode);
+      if (response.headers['content-type']) {
+        res.set('content-type', response.headers['content-type']);
+      }
+      if (response.headers['content-length']) {
+        res.set('content-length', response.headers['content-length']);
+      }
+      if (response.body) {
+        res.send(response.body);
+      }
+    });
+  } catch (err) {
+    console.error(err);
+    return res.status(err.code || 500).send(err.message);
+  }
+};
+// [END email]
+
+// [START verifyWebhook]
+/**
+ * Verify that the webhook request came from sendgrid.
+ *
+ * @param {string} authorization The authorization header of the request, e.g. "Basic ZmdvOhJhcg=="
+ */
+function verifyWebhook (authorization) {
+  var basicAuth = new Buffer(authorization.replace('Basic ', ''), 'base64').toString();
+  var parts = basicAuth.split(':');
+  if (parts[0] !== config.USERNAME || parts[1] !== config.PASSWORD) {
+    var error = new Error('Invalid credentials');
+    error.code = 401;
+    throw error;
+  }
+}
+// [END verifyWebhook]
+
+// [START fixNames]
+/**
+ * Recursively rename properties in to meet BigQuery field name requirements.
+ *
+ * @param {*} obj Value to examine.
+ */
+function fixNames (obj) {
+  if (Array.isArray(obj)) {
+    obj.forEach(fixNames);
+  } else if (obj && typeof obj === 'object') {
+    for (var key in obj) {
+      if (obj.hasOwnProperty(key)) {
+        var value = obj[key];
+        fixNames(value);
+        var fixedKey = key.replace('-', '_');
+        if (fixedKey !== key) {
+          obj[fixedKey] = value;
+          delete obj[key];
+        }
+      }
+    }
+  }
+}
+// [END fixNames]
+
+// [START webhook]
+/**
+ * Receive a webhook from SendGrid.
+ *
+ * See https://sendgrid.com/docs/API_Reference/Webhooks/event.html
+ *
+ * @param {Object} req Cloud Function request context.
+ * @param {Object} res Cloud Function response context.
+ */
+exports.sendgridWebhook = function sendgridWebhook (req, res) {
+  try {
+    if (req.method !== 'POST') {
+      var error = new Error('Only POST requests are accepted');
+      error.code = 405;
+      throw error;
+    }
+
+    verifyWebhook(req.get('authorization') || '');
+
+    var events = req.body || [];
+
+    fixNames(events);
+
+    // Generate newline-delimite JSON
+    // See https://cloud.google.com/bigquery/data-formats#json_format
+    var json = events.map(function (event) {
+      return JSON.stringify(event);
+    }).join('\n');
+
+    // Upload a new file to Cloud Storage if we have events to save
+    if (json.length) {
+      var bucketName = config.EVENT_BUCKET;
+      var filename = uuid.v4() + '.json';
```

Comment:
That's what I meant, unique. Adding timestamps to the filename is a good idea though, I'll do that.